scanpointgenerator/core/compoundgenerator.py

```diff
@@ -76,36 +77,47 @@ def prepare(self):
-                valid &= gen_2.positions[axis_2] \
-                        <= rect.roi.height + rect.roi.start[1]
-                valid &= gen_2.positions[axis_2] >= rect.roi.start[1]
-                points_2 = gen_2.positions[axis_2][valid.astype(np.bool)]
-                new_gen1 = LineGenerator(
-                    gen_1.axes, gen_1.units, points_1[0], points_1[-1],
-                    len(points_1), gen_1.alternate)
-                new_gen2 = LineGenerator(
-                    gen_2.axes, gen_2.units, points_2[0], points_2[-1],
-                    len(points_2), gen_2.alternate)
-                generators[generators.index(gen_1)] = new_gen1
-                generators[generators.index(gen_2)] = new_gen2
-                excluders.remove(rect)
+        for excluder_ in [e for e in excluders if isinstance(e, ROIExcluder)]:
+            rect_rois = [r for r in excluder_.rois
+                         if isinstance(r, RectangularROI) and r.angle == 0]
+            if len(rect_rois) == 1:
+                rect = rect_rois[0]
+                axis_1, axis_2 = excluder_.axes[0], excluder_.axes[1]
+                gen_1 = [g for g in generators if axis_1 in g.axes][0]
+                gen_2 = [g for g in generators if axis_2 in g.axes][0]
+                if gen_1 is gen_2:
+                    continue
+                if isinstance(gen_1, LineGenerator) \
+                        and isinstance(gen_2, LineGenerator):
+                    gen_1.prepare_positions()
+                    gen_2.prepare_positions()
+                    # Filter by axis 1
+                    valid = np.full(gen_1.size, True, dtype=np.int8)
+                    valid &= \
+                        gen_1.positions[axis_1] <= rect.width + rect.start[0]
+                    valid &= \
+                        gen_1.positions[axis_1] >= rect.start[0]
+                    points_1 = gen_1.positions[axis_1][valid.astype(np.bool)]
+                    # Filter by axis 2
+                    valid = np.full(gen_2.size, True, dtype=np.int8)
+                    valid &= \
+                        gen_2.positions[axis_2] <= rect.height + rect.start[1]
+                    valid &= gen_2.positions[axis_2] >= rect.start[1]
+                    points_2 = gen_2.positions[axis_2][valid.astype(np.bool)]
+                    # Recreate generators to replace larger generators + ROI
+                    new_gen1 = LineGenerator(
+                        gen_1.axes, gen_1.units, points_1[0], points_1[-1],
+                        len(points_1), gen_1.alternate)
+                    new_gen2 = LineGenerator(
+                        gen_2.axes, gen_2.units, points_2[0], points_2[-1],
+                        len(points_2), gen_2.alternate)
+                    generators[generators.index(gen_1)] = new_gen1
+                    generators[generators.index(gen_2)] = new_gen2
+                    # Remove ROI from Excluder
+                    excluder_.rois.remove(rect)
```

Comment:
Don't remove the region - it mutates the excluder structure for whoever passed it in (in particular, it invalidates our dict serialization).
It doesn't make any difference inside CompoundGenerator.prepare since the excluder is completely ignored in this case.